HidPkg/UefiHidDxe/src/keyboard.rs

```diff
@@ -8,38 +8,1026 @@
+            }
+          }
+          //Array fields (typically used for key strokes)
+          ReportField::Array(field) => {
+            for usage_list in &field.usage_list {
+              if usage_list.contains(Usage::from(KEYBOARD_USAGE_MIN))
+                || usage_list.contains(Usage::from(KEYBOARD_USAGE_MAX))
+              {
+                report_data.relevant_array_fields.push(ReportFieldWithHandler::<ArrayField> {
+                  field: field.clone(),
+                  report_handler: Self::handle_array_key,
+                });
+                break;
+              }
+            }
+          }
+          ReportField::Padding(_) => (), // padding irrelevant.
+        }
+      }
+      if report_data.relevant_variable_fields.len() > 0 || report_data.relevant_array_fields.len() > 0 {
+        handler.input_reports.insert(report_data.report_id, report_data);
+      }
+    }
+
+    for report in &descriptor.output_reports {
+      let mut report_builder = KeyboardOutputReportBuilder { report_id: report.report_id, ..Default::default() };
+
+      handler.report_id_present = report.report_id.is_some();
+
+      if multiple_reports && !handler.report_id_present {
+        //invalid to have None ReportId if multiple reports present.
+        Err(efi::Status::DEVICE_ERROR)?;
+      }
+
+      report_builder.report_size = report.size_in_bits / 8;
+      if (report.size_in_bits % 8) != 0 {
+        report_builder.report_size += 1;
+      }
+
+      for field in &report.fields {
+        match field {
+          //Variable fields in output reports (typically used for LEDs).
+          ReportField::Variable(field) => {
+            match field.usage.into() {
+              LED_USAGE_MIN..=LED_USAGE_MAX => {
+                report_builder.relevant_variable_fields.push(
+                  ReportFieldBuilder {
+                    field: field.clone(),
+                    field_builder: Self::build_led_report
+                  }
+                )
+              },
+              _=> (), //other usages irrelevant.
+            }
+          },
+          ReportField::Array(_) | // No support for array field report outputs; could be added if required.
+          ReportField::Padding(_) => (), // padding fields irrelevant.
+        }
+      }
+      if report_builder.relevant_variable_fields.len() > 0 {
+        handler.output_builders.push(report_builder);
+      }
+    }
+
+    if handler.input_reports.len() > 0 || handler.output_builders.len() > 0 {
+      Ok(handler)
+    } else {
+      Err(efi::Status::UNSUPPORTED)
+    }
+  }
+
+  //Create the Keyboard Context and install the SimpleTextIn/SimpleTextInEx interfaces.
+  fn install_keyboard_interfaces(
+    self,
+    controller: efi::Handle,
+    hid_context: *mut HidContext,
+  ) -> Result<*mut KeyboardContext, efi::Status> {
+    // retrieve a reference to boot services.
+    // Safety: BOOT_SERVICES must ahve been initialized to point to the UEFI Boot Services table.

```

Comment:
nit typo

```suggestion
    // Safety: BOOT_SERVICES must have been initialized to point to the UEFI Boot Services table.

```

Reply:
I see you fixed it for me - perfect :). Thank you.